cas-server-integration-ehcache/src/main/java/org/jasig/cas/ticket/registry/EhCacheTicketRegistry.java

```diff
@@ -108,12 +110,45 @@ public void addTicket(final Ticket ticket) {
         }
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public boolean deleteTicket(final String ticketId) {
         if (StringUtils.isBlank(ticketId)) {
             return false;
         }
-        return this.serviceTicketsCache.remove(ticketId) || this.ticketGrantingTicketsCache.remove(ticketId);
+
+        final Ticket ticket = getTicket(ticketId);
+        if (ticket == null) {
+            return false;
+        }
+
+        if (ticket instanceof TicketGrantingTicket) {
+            logger.debug("Removing ticket [{}] and its children from the registry.", ticket);
+            return deleteTicketAndChildren((TicketGrantingTicket) ticket);

```

Comment:
I understand the concern about the too many calls. At the same time, I don't think that's a big deal, because the cache is supposed to handle large volumes and loads. The problem with not removing them and simply relying on the expiration policy is that while the ST may be considered expired from the CAS perspective, from the cache perspective it is still present! It is only removed from the cache expiration policy (and not CAS's) kicks in to clean up. That might be too slow, misconfigured and when it kicks in, depending on how GC works it may not actually clean up after itself.

So I think it's a safe bet to do this. Due to referential integrity, if you just remove the TGT from cache, objects that are linked to it may not immediately go away because they may be linked to other objects that are valid. It's a rabbit hole to try to figure out what is linked to what, so removing them all is a safe bet. 

Is there a bulk op with the cache that we could use to delete a batch of objects? If not, I'd be OK with this approach here.